Add generic get regex function

nick-child-ibm opened the pull request:
Dependent on PR #15. I can remove this dependency if it gets held up.

Addresses Issue #4.
This commit adds a function to the `Collector` class for returning a string representation of a regex match when given a command and a regex string. I have demonstrated how this method helps the process of retrieving data by using it in the current data gathering functions.

Let me know if this was not what was intended in the Issue description

erichte-ibm commented on src/collector.rs:
```diff
@@ -51,6 +51,19 @@ impl Collector {
     }
 
+    // returns a string representation of the first capture group of the regex
+    pub fn get_regex_from_command(&mut self, command: &str, regex: &str) -> Option<String> {
```
`parse_from_command`, as it is returning the parsed output, not the regex itself

erichte-ibm replied:
Return a `Result<String, Error>`, so we can use `?` and send the error upwards, or return a value.

erichte-ibm commented on src/collector.rs:
```diff
@@ -51,6 +51,19 @@ impl Collector {
+    pub fn get_regex_from_command(&mut self, command: &str, regex: &str) -> Option<String> {
+        let re = Regex::new(regex).unwrap();
+        let out = self.run_command(command);
+        let output = String::from_utf8(out.stdout).unwrap();
```
some commands (like `gcc -v`) will output on `stderr`, may want to make this an option, or parse both? probably option?

erichte-ibm commented on src/collector.rs:
```diff
@@ -51,6 +51,19 @@ impl Collector {
+        let out = self.run_command(command);
+        let output = String::from_utf8(out.stdout).unwrap();
+        let cap = re.captures(&output).unwrap();
+        if cap.len() < 2 {
+            None
+        } else {
+            Some(String::from(cap.get(1).unwrap().as_str()).clone())
+        }
```
Can we explore this a bit? Perhaps there's an option to let us return the whole cap object here without upsetting the borrow checker, that way we don't have to arbitrarily return only the first.

erichte-ibm commented on src/collector.rs:
```diff
@@ -51,6 +51,19 @@ impl Collector {
 
+    // returns a string representation of the first capture group of the regex
+    pub fn get_regex_from_command(&mut self, command: &str, regex: &str) -> Option<String> {
+        let re = Regex::new(regex).unwrap();
```
This is allowed to panic, our regexes should always at least compile (assuming that's what the possible error is here)

erichte-ibm commented on src/collector.rs:
```diff
@@ -51,6 +51,19 @@ impl Collector {
+        let re = Regex::new(regex).unwrap();
+        let out = self.run_command(command);
+        let output = String::from_utf8(out.stdout).unwrap();
+        let cap = re.captures(&output).unwrap();
```
Return either `None` here if keeping return type as `Option<..>`, or return an `Error` here. We should not panic the whole collection process if one collector dies

erichte-ibm reviewed (changes requested):
1. Don't panic on parse changes
2. Try to return full captures if possible
3. stdout vs stderr

nick-child-ibm commented:
Made following changes:
-changed function name
-changed return type to Result instead of Option
-now we return a Vector<string> of all matches instead of one string of first match
-add parameter for specifying where to read output from stdout or stderr

erichte-ibm commented on src/collector.rs:
```diff
@@ -51,6 +54,27 @@ impl Collector {
+    pub fn parse_from_command(&mut self, command: &str, regex: &str, out_type: OutputStream) -> Result<Vec<String>, CollectorErr> {
+        let re = Regex::new(regex).unwrap();
+        let out = self.run_command(command);
+        let output;
+        match out_type {
+            OutputStream::STDOUT => {output = String::from_utf8(out.stdout)?;}
+            OutputStream::STDERR => {output = String::from_utf8(out.stderr)?;}
+        }
```
```
let output = match out_type { 
           OutputStream::STDOUT => String::from_utf8(out.stdout)?,
           OutputStream::STDERR => output = String::from_utf8(out.stderr)?,
}

erichte-ibm commented on src/collector.rs:
```diff
@@ -51,6 +54,27 @@ impl Collector {
+            OutputStream::STDERR => {output = String::from_utf8(out.stderr)?;}
+        }
+        let cap = re.captures(&output);
+        match cap {
+            None => return Err(format!("Failed to parse output of command '{}' with \
+                                regex '{}' over output:\n{}", command, regex, output).into()),
+            Some(c) => return Ok(
+                                  c.iter()
+                                  .map(| match_str | match_str.unwrap().as_str().to_string())
+                                  .collect()
+                                ),
+        }
```
```
return match cap { ... }
```
